File: tools/sample_models.m
% Author: Ricardo Baptista and Matthias Poloczek
% Date:   June 2018
%
% See LICENSE.md for copyright information
%

function models = sample_models(n_models, n_vars, values)
% SAMPLE_MODELS: Sample candidate model vectors with categorical values
%
% Inputs: n_models - number of vectors to return
%         n_vars - number of variables for each vector
%         values - {n_vars x 1} cell - numeric variables for each dimension

% check values and default to binary if not specified
if nargin < 3
	values = repmat({[0,1]},n_vars,1);
end

% check if n_vars and values have the same dimension
if length(values) ~= n_vars
    error('Number of variables does not match size of values cell')
end

% Generate matrix of zeros to store models
models = zeros(n_models, n_vars);

% Construct model vectors by sampling each dimension independently
% and with replacement
for j=1:n_vars
	models(:,j) = randsample(values{j}, n_models, true);
end

% -- END OF FILE --

File: BOCSpy/sample_models.py
# Author: Ricardo Baptista and Matthias Poloczek
# Date:   June 2018
#
# See LICENSE.md for copyright information
#

import numpy as np

def sample_models(n_models, n_vars):
	# SAMPLE_MODELS: Function samples the binary models to
	# generate observations to train the statistical model

	# Generate matrix of zeros with ones along diagonals
	binary_models = np.zeros((n_models, n_vars))

	# Sample model indices
	model_num = np.random.randint(2**n_vars, size=n_models)

	strformat = '{0:0' + str(n_vars) + 'b}'
	# Construct each binary model vector
	for i in range(n_models):
		model = strformat.format(model_num[i])
		binary_models[i,:] = np.array([int(b) for b in model])

	return binary_models

# -- END OF FILE --